types/index.d.ts

```diff
@@ -295,8 +295,16 @@ declare module 'mongoose' {
     /** Defines a pre hook for the model. */
     pre<T = HydratedDocument<DocType, TInstanceMethods>>(method: 'save', fn: PreSaveMiddlewareFunction<T>): this;
     pre<T = HydratedDocument<DocType, TInstanceMethods>>(method: 'save', options: SchemaPreOptions, fn: PreSaveMiddlewareFunction<T>): this;
+    /**
+     * pre<T = Query<any, any>>(method: MongooseQueryMiddleware | MongooseQueryMiddleware[] | RegExp, options: SchemaPreOptions, fn: PreMiddlewareFunction<T>): this;
+     * with different combinations of SchemaPreOptions.
+     */
+    pre<T = HydratedDocument<DocType, TInstanceMethods>>(method: MongooseQueryMiddleware | MongooseQueryMiddleware[] | RegExp, options:  { document: true, query: false }, fn: PreMiddlewareFunction<T>): this;
```

Comment:
Make `method` explicitly list out the methods where `{ document: true, query: false }` works. So `'remove' | 'updateOne' | 'deleteOne' | 'validate'`. `{ document: true, query: false }` does nothing for `findOneAndUpdate` middleware, for instance.

Reply:
I have added new definition for the middleware method names in order to be able to better distinguish between document only methods, query only methods or ambiguous methods. This led to a long list of cases, and I hope I've got them right (and in the right order).

Note that I had to fix one failing test which wrongly assumed "deleteOnly" to be a query-only middleware method.

We also need to change post accordingly, but we should first clarify the situation with pre only (and copy the results to post later on. That said, I'm afraid that this won't be the last commit.

Reply:
`pre('deleteOne')` registers query middleware, you need to explicitly call `pre('deleteOne', { document: true })` to opt in to getting document middleware for deleteOne. So the test is correct. See the following script:

```javascript
'use strict';

const mongoose = require('mongoose');

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test');

  const schema = mongoose.Schema({ name: String });
  schema.pre('deleteOne', function() {
    console.log('Is query', this instanceof mongoose.Query);
  });

  const Test = mongoose.model('Test', schema);
  const doc = await Test.create({ name: 'foo' });

  console.log('Calling Document.prototype.deleteOne()');
  await doc.deleteOne(); // doesn't print

  console.log('Calling Model.deleteOne()');
  await Test.deleteOne(); // Prints "Is query true"
}

```

Reply:
Of course, you're right. It's so easy to mix things up here... I will fix that, add better annotations for post and, most importantly, will add some tests. Will need some days for that though.

Reply:
@vkarpov15 Are there any TypeScript tests which are actually executed? The tsd-tests are not really executed (but only type checked) -- they only check whether the **inferred** type matches an expected type.
I want to add TypeScript tests which check whether the **static** type matches the **dynamic** type. This is only possible when the test code is executed. I could write plain js tests, but then I won't see the type expectation.

Reply:
I added a plain JS test